core/src/main/java/org/apache/iceberg/io/PartitionedFanoutWriter.java

```diff
@@ -60,11 +93,18 @@ public void write(T row) throws IOException {
 
   @Override
   public void close() throws IOException {
-    if (!writers.isEmpty()) {
-      for (PartitionKey key : writers.keySet()) {
-        writers.get(key).close();
+    ConcurrentMap<PartitionKey, RollingFileWriter> writersMap = writers.asMap();
+    if (writersMap.size() > 0) {
+      // close all remaining rolling file writers
+      try {
+        Tasks.foreach(writersMap.values())
+            .throwFailureWhenFinished()
+            .noRetry()
+            .run(RollingFileWriter::close, IOException.class);
+      } catch (IOException e) {
+        throw new UncheckedIOException("Failed to close rolling file writer", e);
       }
-      writers.clear();
     }
+    writers.invalidateAll();
```

Comment:
Won't this close all of the writers a second time because of the removal listener?

Reply:
Yes I believe it would. The `Tasks.forEach` would call close on the values, but I don't think that Caffeine would know they were closed.

And the removalListener is going to be called for sure on `invalidateAll`. https://github.com/ben-manes/caffeine/wiki/Removal

Reply:
In fact, this is mainly to clear the references in the Map, and I think it has no effect on the second closing of the Writer, because the `BaseTaskWriter` will judge whether the writer is `null`. After the first close, the writer has been set to `null`